makefiles/testing.mak

```diff
@@ -88,6 +88,31 @@ slap: $(CTAGS_TEST)
 		--with-timeout=$(TIMEOUT)"; \
 	$(SHELL) $${c} $(srcdir)/Units
 
+# TODO: Possibly the bmake does not support the metaprogramming similar to gnu make.
```

Comment:
I'd rather avoid using non-portable Make stuff if possible, assuming the rest of our makefiles work with BSD make.  Here it seems fairly easy to simply make it the content of the target itself, it doesn't seem very important to build target, nor even only build once -- rebuilding seems unlikely and innocent enough.

So, I guess I'd just do something like this (not tested at all, just for the argument):
```make
verify-units-inputs: verify-units-inputs-puppet

verify-units-inputs-puppet:
    find $(srcdir)/Units/parser-puppetManifest.r -name expected.tags | $(SED) 's%/[^/]*$$%%' | while read -r dir; do \
        puppet apply --noop "$$dir/input.pp" || exit 1; \
    done

Reply:
I see couple of more problems with this  approach in the comment. 

1) One cannot use the make job control mode (`-j`) for `verify-units-inputs`. All `input.XX` file verification has to be done serially. 
 2) What @b4n mentions: Cannot avoid re-execution. What I have seen, re-execution can be quite expensive. The input language toolchain does not necessarily have fast startup time. In this change, puppet runs take several minutes if executed serially. Consider the need to run interpreters, compilers for each `input.xx` file. Hence, for fast test and development time both (1) and (2) are important. 
2) The complexity of the testing.mak is bound to grow. ctags supports 10's of different input languages. Each having their `verify-untis-inputs-<language_name>` target is not easy to maintain. Compared to that, the conditional of `if !USING_BMAKE ` could be acceptable. 

These three are important drawbacks. 

With using metaprogramming, we gain on all 3 of them. (See my next push to this PR) . What we lose is missing `verify-units-inputs` target in BSD builds. That will not prevent BSD platforms from building and using ctags.  Developers on BSD only won't be able to execute `make verify-units-intput` locally, if they are submitting PRs from BSD platforms. 

I have not done development on BSD before. But that does not seem like a big blocker either https://unix.stackexchange.com/a/127955/212862

Reply:
Fair enough, and you're right that this is probably not a critical enough part of the build system that it can't have its own requirements. Also, we can remember that GNU make is fairly portable by itself, and although not everybody likes it, it is usually possible to get it on most systems; AFAIK it's available on all BSDs under `gmake`, yet probably not in a stock installation.

Anyway, given your arguments I only have one concern now: we should check for GNU make, not against BSD make.  There are *many* Make implementations that aren't compatible with GNU make, and most Make implementations don't have those metaprogramming features.  So unless somebody feels like tesing every single Make implementation around and see what exact subset it supports, I strongly suggests we have a whitelist instead of a blacklist, and only have GNU make in it at first.

Reply:
PS: moreover, most Make implementations are installed as `make`, not as other aliases like `bmake`, `gmake`,  `pmake` or whatnot.  Checking for the name of `$MAKE` is *not* a solid way of cheking which Make implementation is in use.

Reply:
Thanks for the input. I agree on both aspects. The whitelist-blacklist and checking the name of $MAKE. I will hold off on the implementation, though. @masatake mentioned that he will implement this in /misc/units script. If anything changes, I am more than willing to address your comments in this PR.